### ISSUE TITLE
Async comunication with javascript

### ISSUE THREAD

**Issue by jcubic** [kripken/emscripten#6115]
Here is question on SO [How can I run an interactive program compiled with Emscripten in a web page?
](https://stackoverflow.com/q/48292269/387194)

And I came up with a hack that allow to send asynchronous messages from C to javascript and get the value from JS in C. So I can use it with jQuery Terminal read method.

I've abuse the fetch API and create wraper over XHR in JS that send real XHR or my code if url is in special format. The code can be found here 

https://gist.github.com/jcubic/87f2b4c5ef567be43796e179ca08c0da

Does better solution exists, or if you will consider to add this to emscripten to allow for Bidirectional async comunication with javascript.

I was thinking about something like
```c
void read_async();

void success(void *text) {
   printf("input: %s\n", text);
}

void fail() {
   printf("fail\n");
}


void read_async() {
   char** args = {"name: "};
   emscripten_async_js("read", args, &success, &fail);
}

int main() {
  read_async();
}
```

and in JS the code can execute global read function and if returned promise resolve it will execute success in C with data like this:

```javascript
function read(prompt) {
   display(prompt);
   return new Promise(function(resolve, reject) {
      if (true) {
         resolve("hello");
      } else {
         reject();
   });
}
```

and it will execute C function success with string `"hello"`.

**Comment by xloem** [comment]
It looks like a lot of the js libraries used in emterm could be made more general to allow for interoperation with e.g. xterm as a normal terminal.  I started this a little at https://github.com/xloem/emterm .

Emterm lets the user change up their js libraries with `--js-library` .

**Comment by gaycodegal** [comment]
Synchronous terminal communication is possible now thanks to emscripten asyncify.
Code demonstration
https://github.com/gaycodegal/wasm-term

Live demo
https://gaycodegal.github.io/wasm-term/html/

**Comment by TheGeekOnSkates** [comment]
Asyncify is great, but how then would I point fgets or getchar or whatever at it?  I mean I suppose you could pretend you're building an OS from scratch - no standard library, no standard anything - and re-implement these functions as Asyncify-based stdin... but is that really the best way?  If it is, it is.  fgets is out, custom hacks like the OP's read_async or whatever is in.  Forget anything like GNU Readline. :laughing: 

Another thing I've had limited success with is by hacking the "compiled" Javascript and redefining stdin and stdout (a thing I have to redo EVERY TIME I BUILD - so annoying - sorry).  For example, I can redefine console.log to run xterm.js's output function, like I did on one of my [projects](http://www.geekonskates.com/apps/mobile-mainframe/manual/), but stdin is much much harder.  I've tried redefining prompt to return the next character in a buffer, or 0 if there was none, and then use a while-loop in C to wait for input... nope.  It hung the main thread.  So apparently it's Asyncify or bust? :frowning_face: 

There's another option, one I've been trying to avoid but might have to be the way to go: SDL.  SDL is great for games, and has keyboard event handling, so maybe if I geek-rigged a terminal in SDL, maybe *then* I could just use stdin as stdin... Obviously I'd rather not reinvent that wheel, but IMO something really needs to be done to make stdin more usable.  Sorry, I know this sounds an awful lot like a rant, but this single issue has kept me from using Emscripten in any meaningful way since I found it 3-4 years ago.  That's why I'm taking the time to add to this question/issue/whatever.  Thank you for reading, and for all the awesome work you guys do to make WebAssembly accessible to us C programmers. :)

**Comment by gaycodegal** [comment]
Readline is totally something you can emulate in wasm, I do it for my own project. You just need to write your own headers and define functions with the same API as readline. I didn't bother to implement the history part of readline, as that's just extra JavaScript work, but it's totally doable too

**Comment by jackeown** [comment]
Is this possible to do without modifying C code at all? I'd like to turn an arbitrary given interactive C/C++ program into an interactive webpage without requiring any modification of the underlying C code.

**Comment by jcubic** [comment]
@jackeown You can run the interactive code, but I think that you will get `prompt()` in JavaScript when the code uses scanf. If you want a better experience in JavaScript, you need a way to execute async code.

**Comment by jackeown** [comment]
@jcubic  Yeah, but it seems weird (to me) that you wouldn't be able to override that behavior in javascript without modifying the C code.

**Comment by jcubic** [comment]
@jackeown you can't because the C code is synchronous, and alert, prompt, and confirm are the only blocking functions in JavaScript. That's why they were picked by Emscripten. Anything different needs custom C code.
